lua/null-ls/builtins/diagnostics/flake8.lua

```diff
@@ -3,6 +3,48 @@ local methods = require("null-ls.methods")
 
 local DIAGNOSTICS = methods.internal.DIAGNOSTICS
 
+local custom_end_col = {
+    end_col = function(entries, line)
+        local start_col = entries["col"]
+        local message = entries["message"]
+        local code = entries["code"]
+
+        -- highlights only first character on error line, if not specified otherwise
+        local default_position = start_col + 1
+
+        local pattern = nil
+        local trimmed_line = line:sub(start_col, -1)
+
+        if code == "F841" or code == "F823" then
+            pattern = [[local variable %'(.*)%']]
+        elseif code == "F821" or code == "F822" then
+            pattern = [[undefined name %'(.*)%']]
+        elseif code == "F831" then
+            pattern = [[duplicate argument %'(.*)%']]
+        elseif code == "F401" then
+            pattern = [[%'(.*)%' imported]]
+        end
+
+        if pattern == nil then
+            return default_position
+        else
+            local results = message:match(pattern)
+            _, end_col = trimmed_line:find(results, 1, true)
+
+            if end_col == nil then
+                return default_position
+            else
+                end_col = end_col + start_col
+                if end_col > tonumber(start_col) then
+                    return end_col
+                else
+                    return default_position
+                end
+            end
+        end
```

Comment:
```suggestion
        if not pattern then
            return default_position
        end

        local results = message:match(pattern)
        _, end_col = trimmed_line:find(results, 1, true)

        if not end_col then
            return default_position
        end

        end_col = end_col + start_col
        if end_col > tonumber(start_col) then
            return end_col
        end

        return default_position
```

I think using early returns and `not` vs `== nil` is a little more idiomatic and reduces nesting, which will help if this helper continues to grow to handle more codes.